### PR TITLE
feat: improve NetHSM API exception handling

### PR DESCRIPTION
**Pull request by nponsard**
This PR improves the handling of API errors in the nethsm commands.

## Checklist

Make sure to run `make check` and `make fix` before creating a PR, otherwise the CI will fail.

- [x] tested with Python3.9
- [ ] signed commits
- [x] updated documentation (e.g. parameter description, inline doc, docs.nitrokey)
- [x] added labels

## Test Environment and Execution

- OS: Arch
- device's model: NetHSM docker
- device's firmware version: 882a5a3

### CODE REVIEW

File: pynitrokey/nethsm/__init__.py [hunk 1]
```diff
@@ -176,23 +176,49 @@ def __init__(
...
             roles
         )
-    elif e.status == 401 and roles:
-        message = "Unauthorized -- invalid username or password"
+    elif e.status == 405:
+        # 405 "Method Not Allowed" mostly happens when the UserID or KeyID contains a character
+        # - that ends the path of the URL like a question mark '?' :
+        #   /api/v1/keys/?/cert will hit the keys listing endpoint instead of the key/{KeyID}/cert endpoint
+        # - that doesn't cound as a phath parameter like a slash '/' :
+        #   /api/v1/keys///cert will be interpreted as /api/v1/keys/cert with cert as the KeyID
```
Comment (robin-nitrokey) [inline pynitrokey/nethsm/__init__.py:196]: Is that something that we should validate before sending the request?

File: pynitrokey/nethsm/__init__.py [hunk 2]
```diff
@@ -176,23 +176,49 @@ def __init__(
...
         message = "Access denied -- this operation requires the role " + " or ".join(
             roles
         )
-    elif e.status == 401 and roles:
-        message = "Unauthorized -- invalid username or password"
+    elif e.status == 405:
+        # 405 "Method Not Allowed" mostly happens when the UserID or KeyID contains a character
+        # - that ends the path of the URL like a question mark '?' :
+        #   /api/v1/keys/?/cert will hit the keys listing endpoint instead of the key/{KeyID}/cert endpoint
+        # - that doesn't cound as a phath parameter like a slash '/' :
```
Comment (robin-nitrokey) [inline pynitrokey/nethsm/__init__.py:195]: *count *path

File: pynitrokey/nethsm/__init__.py [hunk 3]
```diff
@@ -176,23 +176,49 @@ def __init__(
...
+            "Too many requests -- you may have tried the wrong credentials too often"
+        )
     else:
         message = f"Unexpected API error {e.status}: {e.reason}"
 
     if e.api_response:
         try:
-            body = json.loads(e.api_response.response.data)
+            body = None
+            print(e.api_response)
```
Comment (robin-nitrokey) [inline pynitrokey/nethsm/__init__.py:212]: Debugging artifact?

File: pynitrokey/nethsm/__init__.py [hunk 4]
```diff
@@ -176,23 +176,49 @@ def __init__(
...
+            if hasattr(e.api_response, "text") and e.api_response.text != "":
+                body = json.loads(e.api_response.text)
+            # generated code
+            elif (
+                hasattr(e.api_response, "response")
+                and e.api_response.response.data != ""
+            ):
+                body = json.loads(e.api_response.response.data)
             if "message" in body:
                 message += "\n" + body["message"]
```
Comment (robin-nitrokey) [inline pynitrokey/nethsm/__init__.py:223]: Check if  `body` is `None` (if neither of the two conditions was true).

File: pynitrokey/nethsm/__init__.py [hunk 5]
```diff
@@ -254,9 +280,11 @@ def request(
...
         if not response.ok:
-            e = ApiException(status=response.status_code, reason=response.reason)
-            e.body = response.text
-            e.headers = response.headers
+            e = ApiException(
+                status=response.status_code,
+                reason=response.reason,
+                api_response=response,
+            )
             raise e
```
Comment (robin-nitrokey) [inline pynitrokey/nethsm/__init__.py:288]: nit: remove variable and raise directly?

**Review by robin-nitrokey — COMMENTED** [review]
LGTM, just some comments.

File: pynitrokey/nethsm/__init__.py (hunk 1 above)
Comment (nponsard) [reply pynitrokey/nethsm/__init__.py:196]: That would be the best solution yes

**Review by robin-nitrokey — APPROVED** [review]

File: pynitrokey/nethsm/__init__.py (hunk 1 above)
Comment (robin-nitrokey) [reply pynitrokey/nethsm/__init__.py]: Please create a ticket for it (or a PR ;)).

**Pull request merged**